### PR TITLE
Bundle arcs into a single psfnight job and do similar for flats and cteflats

### PR DESCRIPTION
**Pull request by akremin**
#### Overview
This resolves issues #2371 and #2784 by consolidating all arcs and psfnight jobs into one psfnight job that first runs the individual arc commands and then the joint fit job. A similar thing is done for the flats and nightlyflat, and for the CTE flats.
This is a 6->1, 13->1, and 3->1 reduction in jobs, respectively.

Important choices made:
- I've reduce the number of nodes per arc job to 1 so that all 5 can run in parallel in the Perlmutter realtime queue for daily operations. This is actually more compute-efficient anyway, as noted by Stephen in issue #2784 .
- The twelve flats each need a node and cannot run in parallel within our realtime allocation, so I have throttled them down to 4 nodes. GNU parallel takes care of managing the job execution, which I've tested and works well. This takes three times longer in wall-clock compared to running all in parallel, but is roughly the same in compute time and greatly reduces the job-footprint by a factor of 13.
- I have invented a new cteflat job that bundles the 0-3 CTE flats taken on a night into a single job. This requests one node and each flat is run in serial. There is no joint fit job at the end. If no CTE flats exist on a night, no job is submitted and the code proceeds normally. 

I used coding assistants to help with this PR. I admittedly have not yet reviewed all of the new unit tests written yet, but have reviewed all of the other code.

Another thing that could use some more stress-testing is that all camword corner cases work, but there were multiple dry_run nights that included all exposures with less than 30 cameras and those at least propagate correctly.

Finally, I did not test the ability of the script to pickup if several exposures already exist on disk. The update does check for that and submits only the subset that needs processing, but I didn't explicitly test that functionality as it is unlikely to happen often in practice with calibrations.

#### Tests
My tests are at NERSC: `/global/cfs/cdirs/desi/spectro/redux/jobbundle`

I ran 3 sets of tests:

1. I ran 20260319 in it's entirety from biaspdark through redshifts to verify that the new changes didn't impact other job types. This included 45 biaspdark jobs, which I ran interactively and then restarted the pipeline after some processing table hacking on the ccdcalib job dependencies only. I resubmitted all jobs using `desi_resubmit_queue_failures`, which worked as expected with the new bundled jobs and all downstream jobs. All bundle scripts have the correct number of nodes and parallelism, and the dependencies look right. And the expected output files exist.
2. I ran 20210514 in it's entirety from biaspdark through redshifts to verify that the new changes didn't impact other job types. I only ran with `--n-nights-darks=3` to reduce the number of nights that needed biaspdark jobs. This early night didn't have CTE flats and the pipeline happily submitted the whole night without submitting a cteflat job since there was nothing to do. Again all the scripts look good and outputs exist. As of writing the psfnight job ran and produced the expected outputs, the nightlyflat job is in the queue.
3. In dry_run_level=1 mode, I ran the 16th of every month from 2021-2026 if the exposure_table existed. This produced processing tables and scripts but didn't submit them. All of these produced scripts with the correct number of nodes and  parallelism. 

#### AI disclosure
I used Claude Opus 5 and ChatGPT Sol 5.6 in planning and executing this PR.

### CODE REVIEW

File: py/desispec/workflow/batch_writer.py [hunk 1]
```diff
@@ -549,6 +590,424 @@ def create_ccdcalib_batch_script(night, expids, camword='a0123456789',
...
+    if system_name == 'perlmutter-gpu':
+        mps_wrapper = 'desi_mps_wrapper'
+        step_gpu_opt = '--gpus-per-node={} '.format(batch_config['gpus_per_node'])
+
+    ## OMP_NUM_THREADS is exported once for the whole exposure block
+    if step_class == 'ARC':
+        step_threads = max([res[2] for res in step_resources], default=1)
+    else:
+        step_threads = 1
+
```
[the remark this answers is not in the document]
Comment (sbailey) [reply py/desispec/workflow/batch_writer.py]: I agree with the concern but the proposed fix is incomplete.  I update the code myself to set OMP_NUM_THREADS per arc step.

File: py/desispec/workflow/batch.py [hunk 2]
```diff
@@ -125,6 +125,29 @@ def parse_reservation(reservation, jobdesc):
...
+        int: the largest node count that should be requested for this job type.
+
+    Note:
+        PSFNIGHT is allowed more nodes than everything else because a bundled
+        arc job runs every selected arc exposure concurrently, one node each.
+        Everything else is throttled so that two jobs can run together in the
+        10-node realtime queue.
+    """
+    if jobdesc is not None and jobdesc.upper() == 'PSFNIGHT':
+        return 15
```
Comment (sbailey) [inline py/desispec/workflow/batch.py:146]: This appears to be a leftover from when we thought we'd need 5*3=15 nodes to process arcs in parallel. This branch now uses 1 node per arc exposure for a total of 5 nodes and I don't think this cap of 15 is ever exercised.

File: py/desispec/workflow/batch_writer.py [hunk 3]
```diff
@@ -549,6 +590,421 @@ def create_ccdcalib_batch_script(night, expids, camword='a0123456789',
...
+        queue (str, optional): the Slurm queue that will be used.
+        system_name (str, optional): name of batch system, e.g. perlmutter-cpu.
+
+    Returns:
+        tuple: A tuple containing:
+
+        * nodes: int, number of nodes for this step, always 1.
+        * ntasks: int, number of MPI ranks for this step.
+        * threads_per_task: int, number of threads per rank.
+        * runtime: float, estimated runtime of this step in minutes.
```
Comment (sbailey) [inline py/desispec/workflow/batch_writer.py]: Returning these as a tuple results in other code being concise but cryptic with magic numbers to index the tuple, e.g.
```
    step_runtimes = [res[3] for res in step_resources]
```
I suggest returning these as a dictionary so that they can be accessed with e.g. `res['runtime']` instead of `res[3]`.

**Review by sbailey — COMMENTED** [review]
Overall, looks really good including the testing. I put two questions / suggestions inline in the context of long-term maintainability; they don't impact current functionality.

File: py/desispec/test/test_workflow_batch_writer.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+ARC_NIGHT = 20260706
+ARC_EXPIDS = [360185, 360186, 360187, 360188, 360189]
+FLAT_NIGHT = 20260806
+FLAT_EXPIDS = [364647, 364648, 364649, 364652, 364653, 364654,
+               364657, 364658, 364659, 364662, 364663, 364664]
+CTE_NIGHT = 20260806
+CTE_EXPIDS = [364665, 364666, 364667]
+CTE_CAMWORDS = ['a0123456789', 'a012345678', 'a0123456789']
+FULL_CAMWORD = 'a0123456789'
+REFERENCE_REDUXDIR = '/global/cfs/cdirs/desi/spectro/redux/jobbundle'
```
Comment (sbailey) [inline py/desispec/test/test_workflow_batch_writer.py]: Claude-flavored code leftover: Let's not depend on a hardcoded NERSC path to a test production. Or perhaps this is a string in an evals comparison script that I haven't found, so maybe this just needs an explanatory comment.

File: py/desispec/test/test_workflow_batch_writer.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+
+_evals_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(
+        os.path.dirname(os.path.abspath(__file__))))), 'evals')
```
Comment (sbailey) [inline py/desispec/test/test_workflow_batch_writer.py]: Where is this `evals` dir?  It looks like it is in the top-level checkout of your desispec clone, but never committed?  It would be better to put this under py/desispec/tests/data with a more descriptive name than "evals", and then use `importlib.resources.files` to find it rather than a bunch of chained `dirname` calls.

**Review by sbailey — CHANGES_REQUESTED** [review]
Review part 2, about test infrastructure.

File: py/desispec/workflow/batch.py (hunk 2 above)
Comment (akremin) [reply py/desispec/workflow/batch.py:146]: It is leftover from the original planning of 3 nodes x 5 arc jobs, but I wanted to keep it in case we ever changed our mind in the future. This piece of code is very subtle and has always silently reduced the resources from what was requested based on upper limits we've wanted to enforce. For that reason I wanted the upper limit to be 15 such that we could change to 3x5 nodes in the future if we wanted without extra code changes, while still keeping the sanity check in place so that we don't accidentally code up a 100 node job.

I'm happy to remove the 'psfnight' exception and limit everything to 5 nodes, remove this check entirely and trust the determine_resources() and the people who update it, or leave it as-is with 15 nodes. Please let me know what you'd prefer.

File: py/desispec/test/test_workflow_batch_writer.py (hunk 5 above)
Comment (akremin) [reply py/desispec/test/test_workflow_batch_writer.py]: Yes this was a remnant of my development on my local machine. `evals` was where I put all of the planning and reviewing markdown files and it though those would be committed and kept in the repo, which is not my intension. Sorry for missing this. I'll get this cleaned up

File: py/desispec/workflow/batch.py (hunk 2 above)
Comment (akremin) [reply py/desispec/workflow/batch.py:146]: Speaking to Stephen he is indifferent to the three options listed and left it to me. I will leave it as-is at 15 node maximum. That way if we decide in the future to move back to 3 nodes per arc for speed reasons, we can do so without making this additional change.

File: py/desispec/workflow/batch_writer.py (hunk 3 above)
Comment (akremin) [reply py/desispec/workflow/batch_writer.py]: That's a great suggestion, and I've implemented it in the latest commit.

File: py/desispec/test/test_workflow_batch_writer.py (hunk 4 above)
Comment (akremin) [reply py/desispec/test/test_workflow_batch_writer.py]: This was a remnant of Claude wanting to use my example templates as an end-to-end comparison in unit tests rather than as a design schematic to build toward. I have removed this reference to a file path and the use of templates.

File: py/desispec/test/test_workflow_batch_writer.py (hunk 5 above)
Comment (akremin) [reply py/desispec/test/test_workflow_batch_writer.py]: I have removed the use of templates from the tests and increased the coverage through bespoke tests instead.

File: py/desispec/scripts/proc_night.py [hunk 6]
```diff
@@ -586,13 +586,28 @@ def proc_night(night=None, proc_obstypes=None, z_submit_types=None,
...
+    ## the standard calibrations: a night can have every standard calibration
+    ## accounted for, including a linked fiberflatnight, and still need its own
+    ## cte flats processed for the nightly detector QA.
+    standard_calibs_needed = not all_calibs_submitted(calibjobs['accounted_for'],
+                                                      do_cte_flats, do_darknight)
+    ## cheap pre-check so nights that took no cte flats behave exactly as before
+    night_has_cte_flats = len(split_normal_and_cte_flats(etable)[1]) > 0
     cal_etable = etable[[]]
-    if not all_calibs_submitted(calibjobs['accounted_for'], do_cte_flats, do_darknight):
+    if standard_calibs_needed or (do_cte_flats and night_has_cte_flats):
```
[the remark this answers is not in the document]
Comment (akremin) [reply py/desispec/scripts/proc_night.py:614]: This is fixed in the latest push. It isn't the most important bug since we don't care much about cte flat processing and we care even less about them on nights with no science data, but it was still a bug and worth fixing.

File: py/desispec/workflow/processing.py [hunk 7]
```diff
@@ -407,14 +526,21 @@ def desi_proc_joint_fit_command(prow, queue=None):
...
+    if direct_mode:
+        cmd += ' --mpi'
     return cmd
 
 
 def create_batch_script(prow, queue='realtime', dry_run=0, joint=False,
-                        system_name=None, use_specter=False, extra_job_args=None):
+                        system_name=None, use_specter=False,
+                        check_for_outputs=True, resubmit_partial_complete=True,
+                        extra_job_args=None):
```
[the remark this answers is not in the document]
Comment (akremin) [reply py/desispec/workflow/processing.py:543]: This is not a concern as these are only used within the desispec pipeline infrastructure and are always called by name.

File: py/desispec/scripts/proc_night.py [hunk 8]
```diff
@@ -920,70 +938,92 @@ def submit_calibrations(cal_etable, ptable, cal_override, calibjobs, int_id,
...
+    new_ctes = [cte_erow for cte_erow in ctes
+                if cte_erow['EXPID'] not in processed_cte_expids]
+    if len(new_ctes) > 0:
+        bundle_prow, step_prows, int_id = make_calibration_bundle_prow(
+                new_ctes, descriptor='cteflat', internal_id=int_id,
+                calibjobs=calibjobs)
+        extra_job_args = {'calibration_bundle_steps': step_prows,
+                          'bundle_full_camword': bundle_prow['PROCCAMWORD']}
+        bundle_prow, ptable = create_submit_add_and_save(bundle_prow, ptable,
+                                                         extra_job_args=extra_job_args)
```
[the remark this answers is not in the document]
Comment (akremin) [reply py/desispec/scripts/proc_night.py:984]: Good catch. That was a known limitation that I failed to mention in the PR. Rather than leave it as an undocumented limitation that would require a future update, I've implemented a fix in the latest commit.

**Comment by akremin** [comment]
Tests all pass and all AI and human comments have been resolved. Stephen gave offline approval to merge when ready, so I will proceed with merging
